bluesky_kafka/__init__.py

```diff
@@ -305,3 +305,182 @@ def no_work_during_wait():
+                "do not specify 'group.id' in consumer_config, use only the 'group_id' argument"
+            )
+        else:
+            self._consumer_config["group.id"] = group_id
+
+        if "bootstrap.servers" in self._consumer_config:
+            self._consumer_config["bootstrap.servers"] = ",".join(
+                [bootstrap_servers, self._consumer_config["bootstrap.servers"]]
+            )
+        else:
+            self._consumer_config["bootstrap.servers"] = bootstrap_servers
+
+        logger.info(
+            "starting RemoteDispatcher with Kafka Consumer configuration:\n%s",
+            self._consumer_config,
+        )
+        logger.info("subscribing to Kafka topic(s): %s", topics)
+
+        self._consumer = Consumer(self._consumer_config)
+        self._consumer.subscribe(topics=topics)
+        self.closed = False
+
+    def process_document(self, topic, name, doc):
+        if self._process_document is None:
+            raise NotImplementedError("This class must either be subclassed to override the "
+                                      "process_document method, or have a process function passed "
+                                      "in at init time via the process kwarg.")
+        else:
+            return self._process_document(topic, name, doc)
+
+    def process(self, msg):
+        name, doc = self._deserializer(msg.value())
+        logger.debug(
+            "RemoteDispatcher deserialized document with "
+            "topic %s for Kafka Consumer name: %s doc: %s",
+            msg.topic(),
+            name,
+            doc,
+        )
+        self.process_document(msg.topic(), name, doc)
+
+    def _poll(self, work_during_wait):
+        while True:
+            msg = self._consumer.poll(self.polling_duration)
+            if msg is None:
+                # no message was delivered
+                # do some work before polling again
+                work_during_wait()
+            elif msg.error():
+                logger.error("Kafka Consumer error: %s", msg.error())
+            else:
+                try:
+                    self.process(msg)
+                except Exception as exc:
+                    logger.exception(exc)
+
+    def start(self, work_during_wait=None):
+        def no_work_during_wait():
+            # do nothing between message deliveries
+            pass
+
+        if work_during_wait is None:
+            work_during_wait = no_work_during_wait
+
+        if self.closed:
+            raise RuntimeError(
+                "This RemoteDispatcher has already been "
+                "started and interrupted. Create a fresh "
+                f"instance with {repr(self)}"
+            )
+        try:
+            self._poll(work_during_wait=work_during_wait)
+        except Exception:
+            self.stop()
+            raise
+
+    def stop(self):
+        self._consumer.close()
+        self.closed = True
+
+
+class MongoBlueskyConsumer(BlueskyConsumer):
```

Comment:
Could this class be "BlueskyMongoConsumer" or "MongoConsumer"? "MongoBlueskyConsumer" makes me think there will be classes like "MongoSomethingElseConsumer".

Reply:
Your call on the name, just want to share my thoughts first.  The way I think of the MongoBlueskyConsumer is like Mongo-BlueskyConsumer. Like its a BlueskyConsumer that is specialized for Mongo. I think the name BlueskyConsumer implies that it is a consumer that works on BlueskyDocuments. According to the naming that I had in mind, I suspect that in the future we will probably have a MongoEmbedded-BlueskyConsumer. (I just added the dash to indicate the way I read it.)  I think the MongoEmbedded consumer would inherit from BlueskyConsumer, not MongoBlueskyConsumer. Pick what you like and I'll change the name.

Reply:
My preference would be for the shorter version like "MongoConsumer" as opposed to "MongoBlueskyConsumer". I feel like "Bluesky" is implied and takes too much of the name.

Reply:
Sounds good